File: client/src/main/scala/stasis/client/ops/backup/stages/MetadataCollection.scala
package stasis.client.ops.backup.stages

import java.nio.file.Path
import org.apache.pekko.NotUsed
import org.apache.pekko.stream.scaladsl.Flow
import stasis.client.model.{DatasetMetadata, EntityMetadata, FilesystemMetadata}
import stasis.client.ops.backup.Providers
import stasis.client.tracking.state.BackupState
import stasis.shared.model.datasets.{DatasetDefinition, DatasetEntry}
import stasis.shared.ops.Operation

trait MetadataCollection {
  protected def targetDataset: DatasetDefinition
  protected def latestEntry: Option[DatasetEntry]
  protected def latestMetadata: Option[DatasetMetadata]
  protected def providers: Providers

  def metadataCollection(existingState: Option[BackupState])(implicit
    operation: Operation.Id
  ): Flow[Either[EntityMetadata, EntityMetadata], DatasetMetadata, NotUsed] =
    Flow[Either[EntityMetadata, EntityMetadata]]
      .fold(
        existingState.map(_.asMetadataChanges).getOrElse((Map.empty[Path, EntityMetadata], Map.empty[Path, EntityMetadata]))
      ) { case ((contentChanged, metadataChanged), currentMetadata) =>
        currentMetadata match {
          case Left(metadata)  => (contentChanged + (metadata.path -> metadata), metadataChanged)
          case Right(metadata) => (contentChanged, metadataChanged + (metadata.path -> metadata))
        }
      }
      .map { case (contentChanged, metadataChanged) =>
        DatasetMetadata(
          contentChanged = contentChanged,
          metadataChanged = metadataChanged,
          filesystem = (latestMetadata, latestEntry) match {
            case (Some(metadata), Some(entry)) =>
              metadata.filesystem
                .updated(
                  changes = contentChanged.keys ++ metadataChanged.keys,
                  latestEntry = entry.id
                )

            case _ =>
              FilesystemMetadata(contentChanged.keys ++ metadataChanged.keys)
          }
        )
      }
      .wireTap(_ => providers.track.metadataCollected())
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/backup/stages/MetadataCollection.kt
package stasis.client_android.lib.ops.backup.stages

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.FilesystemMetadata
import stasis.client_android.lib.model.server.datasets.DatasetEntry
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.backup.Providers
import stasis.client_android.lib.tracking.state.BackupState
import stasis.client_android.lib.utils.Either
import stasis.client_android.lib.utils.Either.Left
import stasis.client_android.lib.utils.Either.Right

interface MetadataCollection {
    val latestEntry: DatasetEntry?
    val latestMetadata: DatasetMetadata?
    val providers: Providers

    fun metadataCollection(
        operation: OperationId,
        flow: Flow<Either<EntityMetadata, EntityMetadata>>,
        existingState: BackupState?
    ): Flow<DatasetMetadata> =
        flow {
            val existing = existingState?.asMetadataChanges()

            val contentChanged = existing?.first?.toMutableMap() ?: mutableMapOf()
            val metadataChanged = existing?.second?.toMutableMap() ?: mutableMapOf()

            try {
                flow.collect { currentMetadata ->
                    when (currentMetadata) {
                        is Left -> contentChanged += (currentMetadata.value.path to currentMetadata.value)
                        is Right -> metadataChanged += (currentMetadata.value.path to currentMetadata.value)
                    }
                }
            } finally {
                val metadata = DatasetMetadata(
                    contentChanged = contentChanged,
                    metadataChanged = metadataChanged,
                    filesystem = latestMetadata?.let { metadata ->
                        latestEntry?.let { entry ->
                            metadata.filesystem
                                .updated(
                                    changes = contentChanged.keys + metadataChanged.keys,
                                    latestEntry = entry.id
                                )
                        }
                    } ?: FilesystemMetadata(contentChanged.keys + metadataChanged.keys)
                )

                providers.track.metadataCollected(operation)

                emit(metadata)
            }
        }
}
